### PR TITLE
Issue #4414: Simple name conflict of Checks from thirdparty ch..

### PR DESCRIPTION
**Pull request by Luolc**
#4414

### CODE REVIEW

**Comment by rnveach** [comment]
Failure in wrecker is problem in regression with Apache Struts
> CheckstyleException: In config there is 'HideUtilityClassConstructor', but in classpath there are com.github.sevntu.checkstyle.checks.design.HideUtilityClassConstructorCheck, com.puppycrawl.tools.checkstyle.checks.design.HideUtilityClassConstructorCheck, please resolve ambiguity by specifying fully qualified name in config.

**Comment by romani** [comment]
Wercker failure tell us that if we make behavior like this, all users that have conflict between standard and thirparty will have exception and need to update configuration.

before this PR, we use standard Check, no exceptions.

So we need to make a decision here on behavior:
1) if conflict between standard and thirparty - no exception , use standard. if conflict between thirdparty1 and thirparty2 - exception . [BACKWARD COMPATIBLE, but have uncertain point for future]
2) if conflict between standard and thirparty - exception. if conflict between thirdparty1 and thirparty2 - exception. [BACKWARD IN-COMPATIBLE, fix is easy, very certain/clear behavior for future]

@Luolc , @rnveach , @MEZk , @Vladlis , lets vote. I will share my vote the last to avoid pushing on anybody.

**Comment by Vladlis** [comment]
Second option for sure - failing fast is much better than unclear behavior.

**Comment by rnveach** [comment]
Option 1.
There is no way to remove standard checks as they are embedded into Checkstyle's core. It is always 3rd party's fault that they are using the same, simple name as us.

**Comment by Luolc** [comment]
1) for me.
Before this fix, we are able to use lazy init and hardcode map first to avoid instantiating the third party map. But if we need to throw exception when there are conflicts between standard and thirparty, we would have to instantiate the third party map every time. The instantiating is time-consuming, it would much slow down the test. It would not affect users but does affect development productivity. If we can solve this, then 2) is good.

**Comment by MEZk** [comment]
Point 1.
Agreed with @rnveach . It is not our problem if there is a conflict between 3rd party libs, we should not resolve the conflict, just throw an exception. Checkstyle standard checks in classpath are preferable. Saving backward compatibility is a good solution. Nobody wants to update the tool configuration.

**Comment by romani** [comment]
my vote was for "2)" :).

Looks like "1)" is winning. 3 votes vs 2. 
Ok, lets implement "1)" way.

**Comment by rnveach** [comment]
@Luolc Please make the required changes.
Please rebase over latest master to fix travis issue seen in other PRs.

**Comment by Luolc** [comment]
New implementation for option has been (1) updated.

**Comment by Luolc** [comment]
wercker is building for nearly 12 hours. Strange.

**Comment by rnveach** [comment]
I'm suprised it hasn't timed out by now.
It is stuck on `NoExceptiontest - Hbase` right after the output `Generating "Checkstyle" report`.

**Comment by romani** [comment]
I relaunched wercker.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java [hunk 1]
```diff
@@ -170,38 +188,71 @@ public Object createModule(String name) throws CheckstyleException {
...
+     *      element in {@code fullModuleName}
+     */
+    private Object createObjectFromFullModuleNames(String name, Set<String> fullModuleNames)
+            throws CheckstyleException {
+        if (fullModuleNames.size() == 1) {
+            return createObject(fullModuleNames.iterator().next());
+        }
+        else {
+            final String optionalNames = fullModuleNames.stream()
+                    .reduce((fullNames, fullName) -> fullNames + STRING_SEPARATOR + fullName).get();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: why not simple `.collect(Collectors.joining(", "))` ?

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java [hunk 2]
```diff
@@ -150,6 +165,77 @@ private void testCreateModuleWithNonExistName(String name) {
...
+        final String fooFullName =
+                BASE_PACKAGE + ".packageobjectfactory.foo" + PACKAGE_SEPARATOR + name;
+        final Set<String> fullModuleNames = new HashSet<>(Arrays.asList(barFullName, fooFullName));
+        try {
+            method.invoke(factory, name, fullModuleNames);
+            fail("Exception is expected");
+        }
+        catch (InvocationTargetException ex) {
+            final String optionalNames = fullModuleNames.stream()
+                    .reduce((fullNames, fullName) -> fullNames + STRING_SEPARATOR + fullName).get();
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: please use there concatenation of two string, to clearly show how it should looks like and text main code by different in view test code.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java [hunk 3]
```diff
@@ -170,38 +188,71 @@ public Object createModule(String name) throws CheckstyleException {
...
+     * @param name name of module
+     * @param fullModuleNames the supplied full module names set
+     * @return instance of module if there is only one element in {@code fullModuleName}
+     * @throws CheckstyleException if the class fails to instantiate or there are more than one
+     *      element in {@code fullModuleName}
+     */
+    private Object createObjectFromFullModuleNames(String name, Set<String> fullModuleNames)
+            throws CheckstyleException {
+        if (fullModuleNames.size() == 1) {
+            return createObject(fullModuleNames.iterator().next());
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: please move return to the end of method.

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java [hunk 4]
```diff
@@ -150,6 +165,77 @@ private void testCreateModuleWithNonExistName(String name) {
...
+        final String barPackage = BASE_PACKAGE + ".packageobjectfactory.bar";
+        final String fooPackage = BASE_PACKAGE + ".packageobjectfactory.foo";
+        final PackageObjectFactory packageObjectFactory = new PackageObjectFactory(
+                new HashSet<>(Arrays.asList(barPackage, fooPackage)), classLoader);
+        final Method method = packageObjectFactory.getClass().getDeclaredMethod(
+                "generateThirdPartyNameToFullModuleName", ClassLoader.class);
+        method.setAccessible(true);
+        final Map<String, Set<String>> map =
+                (Map<String, Set<String>>) method.invoke(packageObjectFactory, classLoader);
+        map.values().forEach(fullNames -> assertEquals(2, fullNames.size()));
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: There is nothing asserting that `map` has any keys/values. This test will pass if `map` is completely empty.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java (hunk 1 above)
Comment (Luolc) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java (hunk 2 above)
Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java (hunk 3 above)
Comment (Luolc) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: Done.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties [hunk 5]
```diff
@@ -9,7 +9,11 @@ DefaultLogger.auditStarted=Starting audit...
 DefaultLogger.auditFinished=Audit done.
 
 PackageObjectFactory.unableToInstantiateExceptionMessage=Unable to instantiate ''{0}'' \
-   class, it is also not possible to instantiate it as {1}. \
-   Please recheck that class name is specified as canonical name or read how to configure \
-   short name usage http://checkstyle.sourceforge.net/config.html\#Packages. Please also \
-   recheck that provided ClassLoader to Checker is configured correctly.
+  class, it is also not possible to instantiate it as {1}. \
```
Comment (Luolc) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages.properties]: Done.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties [hunk 6]
```diff
@@ -9,7 +9,10 @@ DefaultLogger.auditStarted=开始检查……
 DefaultLogger.auditFinished=检查完成。
 
 PackageObjectFactory.unableToInstantiateExceptionMessage=无法初始化类： ''{0}'' \
-   ，也无法初始化： {1}。 \
-   请检查类名大小写，或阅读帮助手册中短名称配置部分： \
-   http://checkstyle.sourceforge.net/config.html\#Packages.  \
-   同时，请检查Checker的ClassLoader是否配置正确。
+  ，也无法初始化： {1}。 \
```
Comment (Luolc) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java (hunk 4 above)
Comment (Luolc) [inline src/test/java/com/puppycrawl/tools/checkstyle/PackageObjectFactoryTest.java]: This test is removed now as it is covered by `testCreateObjectFromFullModuleNamesWithException`

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java [hunk 7]
```diff
@@ -170,38 +188,71 @@ public Object createModule(String name) throws CheckstyleException {
 
     /**
      * Create object with the help of the supplied map.
-     * @param name name of module.
-     * @param map the supplied map.
-     * @return instance of module if it is found in modules map.
-     * @throws CheckstyleException if the class fails to instantiate.
+     * @param name name of module
+     * @param map the supplied map
```
Comment (Luolc) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: I see there are no period in other javadoc tag in `PackageObjectFactory`, so I change this to keep them in a same style. If you don't agree with it I would change it back.

**Comment by Luolc** [comment]
@romani 
1. Done
2. I am not very clear how to change the code. Do you mean change the `final ReflectiveOperationException | NoClassDefFoundError ignored` to `ClassNotFoundException ignored`? `Class.forName` throws a checked exception so there have to be a try-catch block.

**Comment by rnveach** [comment]
@Luolc I think he specifically means the `// keep looking, ignoring exception` but we have to keep this ignore when we are trying to instantiate based on full package name at https://github.com/checkstyle/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java#L152-L158 .

File: src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java (hunk 7 above)
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: > see there are no period in other javadoc tag in PackageObjectFactory

I see them in a few places like `PackageObjectFactory.joinPackageNamesWithClassName`. 
Also looking at the history, you added the `map` parameter and the period on the end of that one. :)

Seeing as we have no styles placed on the Javadocs, I don't see removing the periods as necessary and just superfluous as someone else will come around and add them back in.
Please just add the periods back for now. Sometime we will enforce this through a Javadoc style guide.

Comment (Luolc) [inline src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java]: Sure :)

**Comment by Luolc** [comment]
@romani @rnveach I changed to comment to `// ignore the exception` now. Please have a check. :)

**Comment by romani** [comment]
TC is relaunched

**Comment by romani** [comment]
@Luolc ,

>I am not very clear how to change the code

This catch was required when we did blind class loading so expected any exceptions and ignored them. As we do have internal map and dynamic class search , it should be impossible to come cases where class is not accessible, so it should mean serious problem that should be delivered to client as exception. We should have not have logic that is based on exceptions if there is ability to do this by extra IFs.

**Comment by Luolc** [comment]
@romani 
As @rnveach said at https://github.com/checkstyle/checkstyle/pull/4430#issuecomment-308099575, we need change these lines if we want to throw an exception where you mentioned.

Do you think we need to try twice here? Take `ConstantName` as an example, IMO, using `ConstantName`, `ConstantNameCheck`, `com.puppycrawl.tools.checkstyle.checks.naming.ConstantNameCheck` in config is reasonable, but using `com.puppycrawl.tools.checkstyle.checks.naming.ConstantName` is so strange.

And, if we would throw a exception at 
```java
private Object createObject(String className) throws CheckstyleException {
        Class<?> clazz = null;
        try {
            clazz = Class.forName(className, true, moduleClassLoader);
        }
        catch (final ReflectiveOperationException | NoClassDefFoundError ignored) {
            // keep looking, ignoring exception
        }
```
, then 
```java
        if (instance == null) {
            final String attemptedNames = joinPackageNamesWithClassName(name, packages)
                    + STRING_SEPARATOR + nameCheck + STRING_SEPARATOR
                    + joinPackageNamesWithClassName(nameCheck, packages);
            final LocalizedMessage exceptionMessage = new LocalizedMessage(0,
                Definitions.CHECKSTYLE_BUNDLE, UNABLE_TO_INSTANTIATE_EXCEPTION_MESSAGE,
                new String[] {name, attemptedNames}, null, getClass(), null);
            throw new CheckstyleException(exceptionMessage.getMessage());
        }
```
would be never thrown.

And is any suggestion of the exception message? :D

**Comment by romani** [comment]
@Luolc ,

>but using com.puppycrawl.tools.checkstyle.checks.naming.ConstantName is so strange.

Should be an exception in this case, fully qualified name should be absolutely correct.

>would be never thrown.

This code is required, and will be used when we can not find module name in any collections.
Call of `createObject` will be at the end and throw checkstyle wrapped reflection exception in any problem.

**Comment by Luolc** [comment]
@romani
In current code `com.puppycrawl.tools.checkstyle.checks.naming.ConstantName` is valid, see https://github.com/checkstyle/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java#L152-L158. So we are going to remove line 155-158? Please correct me if I misunderstand.

Regarding the exception, I mean, if we throw an exception at https://github.com/checkstyle/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java#L237, then https://github.com/checkstyle/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java#L166 is not reachable. The code would be interrupted at https://github.com/checkstyle/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java#L153, if the name is not exist.

**Comment by romani** [comment]
not finished discussion is moved to separate issue - #4456 to do not allow user to use bad fully qualified names and do refactoring.

**Comment by romani** [comment]
> The code would be interrupted at

extra IF conditions for collection with Check validations are required to be introduced to call `createObject` only then we completely sure in success. Se new issue for details.

**Pull request merged**